### PR TITLE
IGNITE-13320

### PR DESCRIPTION
**Pull request by xtern**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupPageScanner.java [hunk 1]
```diff
@@ -210,21 +210,26 @@ public CacheGroupPageScanner(GridKernalContext ctx) {
             }
 
             Set<Integer> parts = new HashSet<>();
+            long[] pagesLeft = new long[1];
```
Comment (ololo3000) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupPageScanner.java:213]: Why do we need an array here?

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/CacheGroupEncryptionCommand.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    protected abstract String visorTaskName();
+
+    /** Subcommand to Display re-encryption status of the cache group. */
+    protected static class ReencryptionStatus extends
+        CacheGroupEncryptionCommand<Long, VisorCacheGroupEncryptionTaskResult<Long>> {
+        /** {@inheritDoc} */
+        @Override protected void printNodeResult(Long bytesLeft, String grpName, Logger log) {
+            if (bytesLeft == -1)
+                log.info(DOUBLE_INDENT + "re-encryption completed or not required");
+            else
```
Comment (ololo3000) [inline modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/CacheGroupEncryptionCommand.java:139]: New line is redundant.

Comment (xtern) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/CacheGroupEncryptionCommand.java]: Fixed, thanks.

**Review by ololo3000 — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/visor/encryption/VisorCacheGroupEncryptionTaskResult.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    /** */
+    public VisorCacheGroupEncryptionTaskResult() {
+        // No-op.
+    }
+
+    /** @return Per node job result. */
+    public Map<UUID, T> results() {
+        if (results == null)
+            results = new HashMap<>();
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/visor/encryption/VisorCacheGroupEncryptionTaskResult.java:67]: `return results == null ? Collections.emptyMap() : results;`? The same for `exceptions()`

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/ReencryptionRateCommand.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+import static org.apache.ignite.internal.commandline.CommandLogger.INDENT;
+import static org.apache.ignite.internal.commandline.CommandLogger.optional;
+import static org.apache.ignite.internal.commandline.TaskExecutor.BROADCAST_UUID;
+import static org.apache.ignite.internal.commandline.TaskExecutor.executeTaskByNameOnNode;
+import static org.apache.ignite.internal.commandline.encryption.EncryptionSubcommands.REENCRYPTION_RATE;
+
+/**
+ * View/change cache group re-encryption rate limit subcommand.
+ */
+public class ReencryptionRateCommand extends AbstractCommand<VisorReencryptionRateTaskArg> {
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/ReencryptionRateCommand.java:48]: renamed to "reencryption_rate_limit"

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/ReencryptionRateCommand.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                clientCfg
+            );
+
+            boolean read = taskArg.rate() == null;
+
+            Map<UUID, IgniteException> exceptions = res.exceptions();
+
+            for (Map.Entry<UUID, IgniteException> entry : exceptions.entrySet()) {
+                log.info(INDENT + "Node " + entry.getKey() + ":");
+                log.info(String.format("%sfailed to %s re-encryption rate: %s.",
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/ReencryptionRateCommand.java]: Done, thanks

File: modules/core/src/main/java/org/apache/ignite/internal/visor/encryption/VisorCacheGroupEncryptionTaskResult.java (hunk 3 above)
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/visor/encryption/VisorCacheGroupEncryptionTaskResult.java]: Done, thanks

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/ChangeCacheGroupKeyCommand.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.ignite.internal.visor.encryption.VisorChangeCacheGroupKeyTask;
+
+import static org.apache.ignite.internal.commandline.CommandList.ENCRYPTION;
+import static org.apache.ignite.internal.commandline.TaskExecutor.executeTaskByNameOnNode;
+import static org.apache.ignite.internal.commandline.encryption.EncryptionSubcommands.CHANGE_CACHE_GROUP_KEY;
+
+/**
+ * Change cache group key encryption subcommand.
+ */
+public class ChangeCacheGroupKeyCommand extends AbstractCommand<VisorCacheGroupEncryptionTaskArg> {
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/ChangeCacheGroupKeyCommand.java:37]: The CacheGroupEncryptionCommand has been implemented for multinode re-encryption tasks, ChangeCacheGroupKeyCommand is executed on a single server node.

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/ReencryptionRateCommand.java [hunk 7]
```diff
@@ -104,22 +102,13 @@
...
-
-                try {
-                    rateLimit = Double.parseDouble(rateLimitArg);
-                } catch (NumberFormatException e) {
-                    throw new IllegalArgumentException("Failed to parse " + ReencryptionRateCommandArg.LIMIT +
-                        " command argument. Decimal value expected.", e);
-                }
+            try {
+                rateLimit = Double.parseDouble(rateLimitArg);
+            } catch (NumberFormatException e) {
```
Comment (alex-plekhanov) [inline modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/ReencryptionRateCommand.java:117]: NL between `}` and `catch`

Comment (xtern) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/encryption/ReencryptionRateCommand.java]: Done, thanks

**Pull request closed without merging**